overlord/snapstate/handlers.go

```diff
@@ -1333,13 +1333,38 @@ const (
 	unlinkReasonHomeMigration unlinkReason = "home-migration"
 )
 
+// restoreUnlinkOnError assumes that state is locked.
+func (m *SnapManager) restoreUnlinkOnError(t *state.Task, info *snap.Info, tm timings.Measurer) error {
+	st := t.State()
+
+	deviceCtx, err := DeviceCtx(st, t, nil)
+	if err != nil {
+		return err
+	}
+
+	opts, err := SnapServiceOptions(st, info, nil)
+	if err != nil {
+		return err
+	}
+	linkCtx := backend.LinkContext{
+		FirstInstall:   false,
+		IsUndo:         true,
+		ServiceOptions: opts,
+	}
+	_, err = m.backend.LinkSnap(info, deviceCtx, linkCtx, tm)
```

Comment:
double checking: LinkSnap is idempotent, it can be called twice and things work?

Reply:
Yea by my observations, LinkSnap is currently idempotent. I have done a shallow test where I call it 4 times instead of 1 during linking and observe that nothing weird happens, but I'll happily test something more to increase confidence